### PR TITLE
Refactor discrete_model to avoid duplicate digamma calls

### PR DESCRIPTION
**Pull request by jbrockmendel**
Also some flake8 fixes e.g. deleting unused imports.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
I don't have any experience with appveyor.  The error code in the log doesn't help me understand why it failed.  Any ideas?

**Comment by ChadFulton** [comment]
Just a hard-to-debug issue with the state space models, not your code's fault.

File: statsmodels/discrete/discrete_model.py [hunk 1]
```diff
@@ -2073,28 +2069,32 @@ def _score_nbin(self, params, Q=0):
...
+        y = self.endog[:, None]
+        mu = self.predict(params)[:, None]
         a1 = 1/alpha * mu**Q
+        # Note: Q is either 0 or 1, so a1 is either 1/alpha or mu/alpha.
         if Q: # nb1
-            dparams = exog*mu/alpha*(np.log(1/(alpha + 1)) +
-                       special.digamma(y + mu/alpha) -
-                       special.digamma(mu/alpha))
+            digamma_part = (
+                        special.digamma(y+mu/alpha)
```
Comment (josef-pkt) [inline statsmodels/discrete/discrete_model.py]: we follow old pep-8 with spaces around all operators, `+, -, *, /, ...` except `**` (power)

File: statsmodels/discrete/discrete_model.py [hunk 2]
```diff
@@ -2139,14 +2139,16 @@ def _hessian_nb1(self, params):
 
         a1 = mu/alpha
 
+        # for dl/dalpha dalpha
+        digamma_part = (special.digamma(y + mu/alpha) -
+                        special.digamma(mu/alpha))
```
Comment (josef-pkt) [inline statsmodels/discrete/discrete_model.py]: duplicate computation of `mu/alpha`, which is already defined as `a1`

**Comment by josef-pkt** [comment]
looks good as far as I can see from reading through the changes, formulas look unchanged

I made a few comments (which I don't see right now) on smaller things, style, a1

File: statsmodels/discrete/discrete_model.py (hunk 1 above)
Comment (jbrockmendel) [inline statsmodels/discrete/discrete_model.py]: > cosmetic: : I don't like isolated brackets (on separate lines) unless there is a long list of items

[Begrudging acquiescence]

> we follow old pep-8 with spaces around all operators, +, -, *, /, ... except ** (power)

Will fix.

File: statsmodels/discrete/discrete_model.py [hunk 3]
```diff
@@ -2074,10 +2074,8 @@ def _score_nbin(self, params, Q=0):
         a1 = 1/alpha * mu**Q
         # Note: Q is either 0 or 1, so a1 is either 1/alpha or mu/alpha.
         if Q: # nb1
-            digamma_part = (
-                        special.digamma(y+mu/alpha)
-                        - special.digamma(mu/alpha)
-                        )
+            digamma_part = (special.digamma(y + mu / alpha)
+                        - special.digamma(mu / alpha))
```
Comment (josef-pkt) [inline statsmodels/discrete/discrete_model.py]: align below parenthesis, unless we need more space if parenthesis is already far to the right
(next time)

**Comment by josef-pkt** [comment]
the content looks good and ready to merge

However, there is a merge of master in this branch that needs to be removed by an interactive rebase.
`Merge branch 'master'` commit 2e0002e

**Comment by jbrockmendel** [comment]
It turns out other `_score_nbin` and some other methods can be simplified pretty significantly.  I just pushed additional commits for `_score_nbin`.  I broke it up into a bunch of commits, each of which performs ~1 step of algebra, so easy to check for correctness.

**Comment by josef-pkt** [comment]
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-7e6ba889b1c75ec5f3afc889e6dbc58f7c12fa4c)

**fixed alerts:**

* 1 for Module-level cyclic import

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-f2674db5a4126024f63327b9dd6e0278c6571031)

**fixed alerts:**

* 1 for Module-level cyclic import

---

*Comment posted by [lgtm.com](https://lgtm.com)*

**Comment by josef-pkt** [comment]
14 test failures with many nans, so there is either a bug or it is numerically less stable than the old version.

**Comment by jbrockmendel** [comment]
Rebasing is behaving... oddly.  At this point I think it'll be easier to start fresh on a new branch.  Closing.

**Comment by josef-pkt** [comment]
for checking or debugging, you can also compare analytical with numerical derivatives.
In master there is now also a second NegativeBinomialP implementation.

from one of my recent notebooks

```
In [23]:
from statsmodels.tools.numdiff import (approx_fprime, approx_hess,
                                       approx_hess_cs, approx_fprime_cs)
In [24]:
approx_fprime_cs(params, res_nb.model.loglike) 
Out[24]:
array([-1.85303595, -2.30113604,  0.04309051])
In [25]:
res_nb.model.hessian(params)
Out[25]:
array([[ -2.52875864e+01,  -3.07984357e+01,   6.59431360e-01],
       [ -3.07984357e+01,  -4.23212732e+01,   7.89283944e-01],
       [  6.59431360e-01,   7.89283944e-01,  -2.95904325e-02]])
In [26]:
approx_hess_cs(params, res_nb.model.loglike)
Out[26]:
array([[ -2.52875864e+01,  -3.07984356e+01,   6.59431360e-01],
       [ -3.07984356e+01,  -4.23212732e+01,   7.89283945e-01],
       [  6.59431360e-01,   7.89283945e-01,  -2.95904325e-02]])
In [27]:
approx_hess_cs(params, res_nb.model.loglike) - res_nb.model.hessian(params)
Out[27]:
...
```

**Pull request closed without merging**